FlatBuffers post

alexpusch opened the pull request.

Commit:
Start working on FlatBuffers post

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,151 @@
+Case study of Javascript Flat Buffers
+================================
+
+Data serialization is very common process in web application. So common that no one thinks of it. You just pull data from the DB, and your server and client frameworks just handle this thing for you. The server serializes the data into JSON, and the client parses it to JS object. This flow is simple and natural but, as everything, when taken to the extreme start to show scaling issues.
```
the first sentence reads weird..
maybe:
Data serialization is a very common process in web applications

BlueHotDog replied:
I think a lot of the benefit are related to data de-serialization as well.

alexpusch replied:
👍

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,151 @@
+
+Whats an 'extreme' for data serialization? A lot of data. Serializing and deseriazling tens of thousands of objects takes it toll from the server, and more importantly, from the single threaded browser page.
+
+A possible solution for this scaling issue are FlatBuffers. We've utilized this tool to do our heavy lifting and lived to tell about it.
```
is Flatbuffers? it's a single technology

alexpusch replied:
👍

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,151 @@
+
+This undertaking raises many challenges, one of which is to transfer all of those events from the recorded user to our servers, and from our servers to the player. The efficiency of this process is measured both by the size of the transfered data and by the time it takes to parse it, measures that JSON does not excel in.
+
+# What are FlatBuffers?
```
Flatbuffer vs other solutions?

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,151 @@
+Case study of Javascript Flat Buffers
```
seems this is more of a flatbuffer vs json post and should gain more traffic

alexpusch replied:
You're right, but the post talks about our specific experience. What about "Refactoring JSON to FlatBuffers", "Moving from JSON to FlatBuffers", "Case study of JSON to FlatBuffers transition". Any ideas?

itayadler replied:
I like Refactoring JSON to FlatBuffers

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
````diff
@@ -0,0 +1,151 @@
+root_type Event;
+```
+
+After defining the schema, we run the flatbuffer 'compiler' - flatc. This generates code in our language of choice. That code handles all the bytes and bits calculation flatbuffer is based on.
````
one thing to mention is that flatbuffer are available to a LOT of languages aside from js

alexpusch replied:
I've added it to the introduction

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,151 @@
+
+All this data must be stored and accessible for the recorders counterpart - the player. The player knows how to take the mass of events and reconstruct them into a watchable clip.
+
+This undertaking raises many challenges, one of which is to transfer all of those events from the recorded user to our servers, and from our servers to the player. The efficiency of this process is measured both by the size of the transfered data and by the time it takes to parse it, measures that JSON does not excel in.
```
I would also add a background check we did of FB vs other similar frameworks we did and why we chose flatbuffers..

alexpusch replied:
did we do that? what were the conclusions?

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
````diff
@@ -0,0 +1,151 @@
+console.log('event data', event.event_data(new EventSerializer.MousePositionData())).x()
+```
+
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
````
maybe talk about actual cons? like no typesafe, building is form the outside in

alexpusch replied:
good point. I have a drawbacks section at the bottom, I'll add these points

BlueHotDog commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,151 @@
+* Non human readable
+* Verbal and cumbersome API
+* Backwards compatibility issues
+* Lack of streaming support
```
I would write a conclusion
And also touch put in the benchmarks speed comparison.

alexpusch replied:
👍

BlueHotDog reviewed (commented):
very awesome!

Commit:
Added drawbecks, conclusions, and fixed other stuff

- turns out the space benckmark result I've put in the previous commit
were really good to be true :(

Commit:
Added example code

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
+
+# Phases of refactor
+Refactoring our entire flow from JSON to FlatBuffers is quite a feat. It's a architecture crossing change that will be dangerous to try and accomplish in one go. We've considered this issue for a while and managed to split the redactor into several independent steps.
```
redactor => refactor

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+
+In addition to the parsing efficiency the binary format gains us another advantage - binary representation of data is often more size efficient. We can store a 10 digit int using 4 bytes UInt, instead of 10 chars.
+
+## Flat buffers Example
```
Flat buffers => FlatBuffers
I think this should be consistent in the entire article

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+# What are FlatBuffers?
+[FlatBuffers](https://google.github.io/flatbuffers) is a serialization library by Google. It was first made by Google with android games in mind, but now it has ports to C++, C#, C, Go, Java, PHP, Python, and JavaScript.
+
+The main goal of FlatBuffers is to avoid deseriazation what so ever. This is achieved by defining a binary data protocol - a well defined method to convert data to, and from binary. The binary structure created by this protocol can be sent over the wire, and read without further processing. To reiterate, when transferring JSON, we need to transform data into a string, send it over the wire, parse ...
```
deseriazation => deserialization

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+In addition to the parsing efficiency the binary format gains us another advantage - binary representation of data is often more size efficient. We can store a 10 digit int using 4 bytes UInt, instead of 10 chars.
+
+## Flat buffers Example
+This won't be a FlatBuffers tutorial, but to give you a feel for what it's like lets took on a simplified version of our data model.
```
took on => look at

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
````diff
@@ -0,0 +1,172 @@
+root_type Event;
+```
+
+After defining the schema, we run the flatbuffer 'compiler' - flatc. This generates code in our language of choice. That code handles all the bytes and bits calculation flatbuffer is based on.
````
flatbuffer => FlatBuffers

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
````diff
@@ -0,0 +1,172 @@
+console.log('event data', event.event_data(new EventSerializer.MousePositionData())).x()
+```
+
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
````
Flatbuffers => FlatBuffers

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+Refactoring our entire flow from JSON to FlatBuffers is quite a feat. It's a architecture crossing change that will be dangerous to try and accomplish in one go. We've considered this issue for a while and managed to split the redactor into several independent steps.
+
+## FlatBuffers in recorder
+We began in transforming our recorder messages to FlatBuffers. The recorder creates and send the recorded events as FlatBuffers. The first step of the processing pipeline is to transform the binary data back to JSON allowing the rest of the architecture to stay unchanged.
```
send => seds

alexpusch replied:
done

itayadler replied:
bah seds :) I mean't sends

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+In fact we have not taken this step yet. A Major benefit of JSON is that it's easily streamable using libraries such as [Oboe.js](http://oboejs.com/). We have yet to discover a method to achieve a similar goal using FlatBuffers.
+
+# benchmarks
+## Space banchmark
```
Space banchmark => Storage size benchmark

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+The result did not met our expectations, not only the FlatBuffers file are not smaller that our JSON, they are 15%-30% bigger. While this is a disappointing result the overall file size is just a secondary goal. This issue will definitely be studies.
+
+## Data availability speed benchmark
+The next important benchmark would be data availability speed. By this I mean the duration of time it takes from initial request to complete data availability. In this benchmark we will take several session, in both JSON and FlatBuffers format, download and iterate over their data to perform some simple calculation.
```
several session => several sessions

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+
+<some kind of graph>
+
+As expected, JSON is impeded by the parsing stage. FlatBuffers allows immediate access to data after download. Those several seconds of JSON parsing are critical for us. These are the first seconds of our page loads, where the first impression of our system takes place.
```
`These are the first seconds of our page loads, where the first impression of our system takes place.`
something is off with this sentence

alexpusch replied:
how do you think it should be?

itayadler replied:
Those several seconds of JSON parsing are critical for us. These are the first seconds of our page loads, where the first impression of our system takes place. 
=>
The time spent on JSON parsing during page load time is noticeable when dealing with a higher volume of data, and if we can shave it off, then that'll be key to delivering a non-janky experience to our users.

What do you think?

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+
+* Lack of streaming support - When dealing with large amount of data there is a good chance you would like to stream it. There are several issues with streaming a FlatBuffer array. Flatbuffers are written backwards. This means that crucial bits of our data appear at the end of the file, making streaming not feasible. Even if this issue could be overcome another hurdle exists. XmlHttpRequest does  ...
+
+# Conclusion
```
we pretty pleased => we are pretty pleased

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+# Conclusion
+Several months into the FlatBuffers refactor we pretty pleased with the result. We've seen a noticeable performance improvement in the areas of the architecture we've converted to FlatBuffers.
+
+This was not an easy ride. There are several caveats in FlatBuffers that gets exaggerated by JavaScatips weak typing nature.
```
JavaScatips => Javacript's

alexpusch replied:
done

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -0,0 +1,172 @@
+
+This was not an easy ride. There are several caveats in FlatBuffers that gets exaggerated by JavaScatips weak typing nature.
+
+FlatBuffer is heavy machinery. It won't do you any good if you don't have a vast number of objects to throw at it.
```
FlatBuffer => FlatBuffers ?

alexpusch replied:
done

Commit:
PR

itayadler commented on flatbuffers/flatbuffers_case_study.md:
```diff
@@ -138,19 +138,19 @@ The Final step is to actually store the FlatBuffers bytes in our storage. This i
 
 # benchmarks
-## Space banchmark
+## storage benchmark
```
Captial S
